### PR TITLE
Update blame links with selected line, fix bug 1007293

### PR DESCRIPTION
**Pull request by pelmers**
take 2! This time simplified from #448 (no format bump, mostly a javascript update).

Plugins can put {{line}} in the href of any links, and these will be substituted with the most recently selected line number.

### CODE REVIEW

**Comment by pelmers** [comment]
It's all javascript!
r? @mythmon

File: dxr/static_unhashed/js/code-highlighter.js [hunk 1]
```diff
@@ -132,6 +138,26 @@ $(function () {
...
+        navlinks.each(function() {
+            var $this = $(this);
+            // Copy the href to a template attr, since we will be updating it
+            // and don't want to lose the template. This behavior will happen
+            // only the first time this function is called.
+            if (!$this.data('template')) {
+                $this.data('template', $this.attr('href'));
+            }
+            if (lastNumber) {
+                $this.attr('href', $this.data('template').replace(/{{line}}/g, lastNumber));
```
Comment (mythmon) [inline dxr/static_unhashed/js/code-highlighter.js]: Does this work? I know that `{}` are special characters in some regex dialects. Not sure if it is a problem here.

File: dxr/vcs.py [hunk 2]
```diff
@@ -147,7 +147,7 @@ def generate_diff(self, path):
         return self.upstream + 'diff/' + self.previous_revisions[path] + '/' + path
 
     def generate_blame(self, path):
-        return self.upstream + 'annotate/' + self.revision + '/' + path
+        return self.upstream + 'annotate/' + self.revision + '/' + path + '#l{{line}}'
```
Comment (mythmon) [inline dxr/vcs.py]: Probably makes sense to use string formatting here, instead of repeated concatenation.

**Comment by mythmon** [comment]
This seems like a fragile approach, since if the JS screws up you would get bad urls in your links. Maybe have the thing that generates the HTML fill in `data-template` directly with the template, and `href` with a safe fallback? That also simplifies the JS a bit.

**Comment by pelmers** [comment]
Yes it would! I'll do that and ping you again

File: dxr/static_unhashed/js/code-highlighter.js (hunk 1 above)
Comment (pelmers) [inline dxr/static_unhashed/js/code-highlighter.js]: It did work when I tested it locally. I think curlies are only used in {n} to match n occurrences of the preceding character.

File: dxr/static_unhashed/js/code-highlighter.js [hunk 3]
```diff
@@ -143,17 +143,13 @@ $(function () {
...
-                $this.attr('href', $this.data('template').replace(/{{line}}/g, lastNumber));
-            } else {
-                $this.attr('href', $this.data('template').replace(/{{line}}/g, ''));
+            const $this = $(this);
+            if ($this.data('template')) {
+                if (lastNumber) {
+                    $this.attr('href', $this.data('template').replace(/{{line}}/g, lastNumber));
+                } else {
+                    $this.attr('href', $this.data('template').replace(/{{line}}/g, ''));
+                }
```
Comment (mythmon) [inline dxr/static_unhashed/js/code-highlighter.js]: You can be a bit more concise here like this:

```javascript
$this.attr('href', $this.data('template').replace(/{{line}}/g, lastNumber || ''));
```

**Comment by pelmers** [comment]
@mythmon thanks for the reviews!

File: dxr/static_unhashed/js/code-highlighter.js [hunk 4]
```diff
@@ -327,7 +345,9 @@ $(function () {
     }
 
     // Highlight any lines specified by hash in either a direct page load or a history pop.
-    $(document).ready(processHash);
+    $(document).ready(function() {
```
Comment (erikrose) [inline dxr/static_unhashed/js/code-highlighter.js]: Why wrap this in a function? The terser original form was equivalent, no?

File: dxr/templates/file.html [hunk 5]
```diff
@@ -37,7 +37,11 @@
           <ul>
             {% for item in section['items'] %}
               <li>
-                <a href="{{ item.href }}" title="{{ item.title }}" class="{{ item.icon }} icon">{{ item.title }}</a>
+                {% if 'template' in item %}
```
Comment (erikrose) [inline dxr/templates/file.html]: I have a slight preference for DRYing this and putting the `if` only around the template attr.

File: dxr/indexers.py [hunk 6]
```diff
@@ -224,6 +224,9 @@ def links(self):
 
             (sort order, heading, [(icon, title, href), ...])
 
+            File views will replace any {{line}} within the href with the
+            last-selected line number.
```
Comment (erikrose) [inline dxr/indexers.py]: So we scroll to the first-selected line but blame to the last. Any reason for the inconsistency?

Comment (pelmers) [inline dxr/indexers.py]: No, actually when I was writing this I hadn't realized that we scrolled to the first selected line.

**Comment by erikrose** [comment]
My only reservation is that there's a hanging `#L` on any blame URLs that don't have a line number. It shouldn't hurt anything; it just looks shoddy. But it doesn't need to block merging this. r+. (And file a bug about the trailing #L.)

**Comment by pelmers** [comment]
Thanks for review!

**Pull request merged**